src/Optimizers/ReuseComponents.ts

```diff
@@ -0,0 +1,55 @@
+import { ReportElement } from '../Models/Report';
+import { AsyncAPIDocument } from '@asyncapi/parser';
+import { ComponentProvider } from '../ComponentProvider';
+import { compareComponents } from '../Utils';
+import { OptimizerInterface } from '../Models/OptimizerInterface';
+
+/**
+ * This optimizer will find all of the components that are declared in _components_ section of the AsyncAPI spec that can be reused in other part of the spec and generate a detailed report of them.
+ *
+ * @public
+ */
+export class ReuseComponents implements OptimizerInterface {
+  provider: ComponentProvider;
+
+  constructor(private document: AsyncAPIDocument) {
+    this.provider = ComponentProvider.getInstance(document);
+  }
+
+  /**
+   * After initializing this class, getReport function can be used to generate a report of components that can be reused.
+   *
+   * @returns ReportElement[]
+   * @defaultValue `true`
+   */
+  getReport = (): ReportElement[] => {
+    return this.findDuplicateComponents(this.provider.schemas).concat(
+      this.findDuplicateComponents(this.provider.messages),
+      this.findDuplicateComponents(this.provider.parameters)
+    );
+  }
+
+  findDuplicateComponents = (component: Map<string, any>): ReportElement[] => {
+    const elements = [];
+    for (const [key1, value1] of component) {
+      for (const [key2, value2] of component) {
+        if (key1 === key2 || !this.isChannelToComponent(key1, key2)) {
+          continue;
+        }
+        if (value1.json() !== value2.json() && compareComponents(value1.json(), value2.json())) {
```

Comment:
One a first glance, this line is not conveying what it is doing.

Simply translating, it seems like:
If JSON of value1 and JSON of value2 are not equal (it is fine till here) and compare components value1 and value2

Instead a better naming would have been somthing like `areSame(value1.json(), value2.json())

Or even better if value1 is an object of some class, you can have some method on it
```js
class Component {
   isEqual(component) {
   // Logic here
  }
}
```

when then can be used as 
```js
value1.isEqual(value2)
```

Reply:
Looking at other places of this, the entire line can be a class method like

```js
class Component {
  isEqual(value) {
     return this.json() !== value.json() && compareComponents(this.json(), value.json())
  }
}
```